## 1. Importing the necessary packages

In [1]:
# Importing the necessary function
import pandas as pd
import numpy as np
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score,classification_report,confusion_matrix

In [2]:
#Importing & defining the datframe
df=pd.read_csv('dataset_35_dermatology.csv')
df.head()

erythema  scaling  definite_borders  itching  koebner_phenomenon  \
0         2        2                 0        3                   0   
1         3        3                 3        2                   1   
2         2        1                 2        3                   1   
3         2        2                 2        0                   0   
4         2        3                 2        2                   2   

   polygonal_papules  follicular_papules  oral_mucosal_involvement  \
0                  0                   0                         0   
1                  0                   0                         0   
2                  3                   0                         3   
3                  0                   0                         0   
4                  2                   0                         2   

   knee_and_elbow_involvement  scalp_involvement  ...  \
0                           1                  0  ...   
1                           1                  1  ...   
2                           0                  0  ...   
3                           3                  2  ...   
4                           0                  0  ...   

   disappearance_of_the_granular_layer  \
0                                    0   
1                                    0   
2                                    0   
3                                    3   
4                                    2   

   vacuolisation_and_damage_of_basal_layer  spongiosis  \
0                                        0           3   
1                                        0           0   
2                                        2           3   
3                                        0           0   
4                                        3           2   

   saw-tooth_appearance_of_retes  follicular_horn_plug  \
0                              0                     0   
1                              0                     0   
2                              2                     0   
3                              0                     0   
4                              3                     0   

   perifollicular_parakeratosis  inflammatory_monoluclear_inflitrate  \
0                             0                                    1   
1                             0                                    1   
2                             0                                    2   
3                             0                                    3   
4                             0                                    2   

   band-like_infiltrate  Age  class  
0                     0   55      2  
1                     0    8      1  
2                     3   26      3  
3                     0   40      1  
4                     3   45      3  

[5 rows x 35 columns]

## 3. Checking the null Values

In [3]:
df.replace('?',np.nan,inplace=True)
df.head(5)

erythema  scaling  definite_borders  itching  koebner_phenomenon  \
0         2        2                 0        3                   0   
1         3        3                 3        2                   1   
2         2        1                 2        3                   1   
3         2        2                 2        0                   0   
4         2        3                 2        2                   2   

   polygonal_papules  follicular_papules  oral_mucosal_involvement  \
0                  0                   0                         0   
1                  0                   0                         0   
2                  3                   0                         3   
3                  0                   0                         0   
4                  2                   0                         2   

   knee_and_elbow_involvement  scalp_involvement  ...  \
0                           1                  0  ...   
1                           1                  1  ...   
2                           0                  0  ...   
3                           3                  2  ...   
4                           0                  0  ...   

   disappearance_of_the_granular_layer  \
0                                    0   
1                                    0   
2                                    0   
3                                    3   
4                                    2   

   vacuolisation_and_damage_of_basal_layer  spongiosis  \
0                                        0           3   
1                                        0           0   
2                                        2           3   
3                                        0           0   
4                                        3           2   

   saw-tooth_appearance_of_retes  follicular_horn_plug  \
0                              0                     0   
1                              0                     0   
2                              2                     0   
3                              0                     0   
4                              3                     0   

   perifollicular_parakeratosis  inflammatory_monoluclear_inflitrate  \
0                             0                                    1   
1                             0                                    1   
2                             0                                    2   
3                             0                                    3   
4                             0                                    2   

   band-like_infiltrate  Age  class  
0                     0   55      2  
1                     0    8      1  
2                     3   26      3  
3                     0   40      1  
4                     3   45      3  

[5 rows x 35 columns]

## 4 Replacing The nulll values with mean

In [4]:
columns = ['Age']
for col in columns:
    df[col]=pd.to_numeric(df[col])
    df[col].fillna(df[col].mean(), inplace=True)
df.head() 

erythema  scaling  definite_borders  itching  koebner_phenomenon  \
0         2        2                 0        3                   0   
1         3        3                 3        2                   1   
2         2        1                 2        3                   1   
3         2        2                 2        0                   0   
4         2        3                 2        2                   2   

   polygonal_papules  follicular_papules  oral_mucosal_involvement  \
0                  0                   0                         0   
1                  0                   0                         0   
2                  3                   0                         3   
3                  0                   0                         0   
4                  2                   0                         2   

   knee_and_elbow_involvement  scalp_involvement  ...  \
0                           1                  0  ...   
1                           1                  1  ...   
2                           0                  0  ...   
3                           3                  2  ...   
4                           0                  0  ...   

   disappearance_of_the_granular_layer  \
0                                    0   
1                                    0   
2                                    0   
3                                    3   
4                                    2   

   vacuolisation_and_damage_of_basal_layer  spongiosis  \
0                                        0           3   
1                                        0           0   
2                                        2           3   
3                                        0           0   
4                                        3           2   

   saw-tooth_appearance_of_retes  follicular_horn_plug  \
0                              0                     0   
1                              0                     0   
2                              2                     0   
3                              0                     0   
4                              3                     0   

   perifollicular_parakeratosis  inflammatory_monoluclear_inflitrate  \
0                             0                                    1   
1                             0                                    1   
2                             0                                    2   
3                             0                                    3   
4                             0                                    2   

   band-like_infiltrate   Age  class  
0                     0  55.0      2  
1                     0   8.0      1  
2                     3  26.0      3  
3                     0  40.0      1  
4                     3  45.0      3  

[5 rows x 35 columns]

In [5]:
X = df.iloc[:,:34]
X.head()

erythema  scaling  definite_borders  itching  koebner_phenomenon  \
0         2        2                 0        3                   0   
1         3        3                 3        2                   1   
2         2        1                 2        3                   1   
3         2        2                 2        0                   0   
4         2        3                 2        2                   2   

   polygonal_papules  follicular_papules  oral_mucosal_involvement  \
0                  0                   0                         0   
1                  0                   0                         0   
2                  3                   0                         3   
3                  0                   0                         0   
4                  2                   0                         2   

   knee_and_elbow_involvement  scalp_involvement  ...  focal_hypergranulosis  \
0                           1                  0  ...                      0   
1                           1                  1  ...                      0   
2                           0                  0  ...                      2   
3                           3                  2  ...                      0   
4                           0                  0  ...                      2   

   disappearance_of_the_granular_layer  \
0                                    0   
1                                    0   
2                                    0   
3                                    3   
4                                    2   

   vacuolisation_and_damage_of_basal_layer  spongiosis  \
0                                        0           3   
1                                        0           0   
2                                        2           3   
3                                        0           0   
4                                        3           2   

   saw-tooth_appearance_of_retes  follicular_horn_plug  \
0                              0                     0   
1                              0                     0   
2                              2                     0   
3                              0                     0   
4                              3                     0   

   perifollicular_parakeratosis  inflammatory_monoluclear_inflitrate  \
0                             0                                    1   
1                             0                                    1   
2                             0                                    2   
3                             0                                    3   
4                             0                                    2   

   band-like_infiltrate   Age  
0                     0  55.0  
1                     0   8.0  
2                     3  26.0  
3                     0  40.0  
4                     3  45.0  

[5 rows x 34 columns]

## 5. Feature Selection with PCA

In [6]:
from sklearn.decomposition import PCA
pca = PCA()
x=pd.DataFrame(pca.fit_transform(X))

In [7]:
x.head()

0         1         2         3         4         5         6   \
0  18.608512  1.305720 -3.367014 -0.737118 -1.614683  0.302348  0.731256   
1 -28.233465 -2.603527  2.348490 -0.178426 -1.564858 -0.411541 -0.045377   
2 -10.222370  4.894243  2.566681  0.090485 -0.621563 -0.882856  0.202208   
3   3.755951 -4.415770  2.150404 -1.501499  0.603335 -1.105585  0.222283   
4   8.788378  4.513829  2.477289 -0.518925  0.184211 -0.053230 -1.187233   

         7         8         9   ...        24        25        26        27  \
0 -0.910474  0.866924 -0.754471  ... -0.111838  0.330791  0.065691 -0.077955   
1  1.195887 -1.353300 -0.508558  ... -0.284200  0.081901 -0.503003  0.466846   
2 -0.586878  0.039572  1.011984  ... -0.140815  0.482840  0.528838  0.545843   
3  1.065908  1.829714 -0.138918  ... -0.360179  0.188301  0.012150 -0.097563   
4 -0.128757  0.158732 -0.998945  ...  0.539760  0.037983 -0.921167 -0.110463   

         28        29        30        31        32        33  
0 -0.077010 -0.086145 -0.110463 -0.067583  0.001793  0.027045  
1 -0.256010  0.095762  0.098145 -0.159200  0.009671 -0.029855  
2  1.073597 -0.064357  0.164594  0.005684  0.369622  0.030755  
3  0.117900  0.080074  0.029023  0.018303  0.043345  0.052816  
4  0.307305 -0.486680  0.420600 -0.025680  0.884592 -0.008932  

[5 rows x 34 columns]

## 5. Defining the 'X' and 'y'

In [8]:
x = x.iloc[:,:10]
y=df['class']

In [9]:
x.head()

0         1         2         3         4         5         6  \
0  18.608512  1.305720 -3.367014 -0.737118 -1.614683  0.302348  0.731256   
1 -28.233465 -2.603527  2.348490 -0.178426 -1.564858 -0.411541 -0.045377   
2 -10.222370  4.894243  2.566681  0.090485 -0.621563 -0.882856  0.202208   
3   3.755951 -4.415770  2.150404 -1.501499  0.603335 -1.105585  0.222283   
4   8.788378  4.513829  2.477289 -0.518925  0.184211 -0.053230 -1.187233   

          7         8         9  
0 -0.910474  0.866924 -0.754471  
1  1.195887 -1.353300 -0.508558  
2 -0.586878  0.039572  1.011984  
3  1.065908  1.829714 -0.138918  
4 -0.128757  0.158732 -0.998945

In [10]:
y.head()

0    2
1    1
2    3
3    1
4    3
Name: class, dtype: int64

## 6. Model-4 => RandomForest Model

In [11]:
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=25)

In [12]:
from sklearn.ensemble import RandomForestClassifier
model=RandomForestClassifier( n_estimators=100,max_depth=15,random_state=4)

In [13]:
model.fit(x_train,y_train)
y_predict=model.predict(x_test)

## 7. Model Evaluation

In [14]:
accuracy_score(y_test,y_predict)

0.9459459459459459

In [15]:
print(classification_report(y_test,y_predict))

              precision    recall  f1-score   support

           1       1.00      1.00      1.00        21
           2       0.94      0.89      0.91        18
           3       1.00      1.00      1.00        15
           4       0.70      0.88      0.78         8
           5       1.00      0.89      0.94         9
           6       1.00      1.00      1.00         3

    accuracy                           0.95        74
   macro avg       0.94      0.94      0.94        74
weighted avg       0.95      0.95      0.95        74



In [16]:
confusion_matrix(y_test,y_predict)

array([[21,  0,  0,  0,  0,  0],
       [ 0, 16,  0,  2,  0,  0],
       [ 0,  0, 15,  0,  0,  0],
       [ 0,  1,  0,  7,  0,  0],
       [ 0,  0,  0,  1,  8,  0],
       [ 0,  0,  0,  0,  0,  3]], dtype=int64)

## 8. Model Evaluation with Cross validation

In [17]:
from sklearn.model_selection import cross_val_score
scores = cross_val_score(model,X,y,cv=15,scoring="accuracy")
print(scores)

[1.         0.92       0.92       1.         1.         1.
 1.         0.95833333 1.         0.95833333 1.         0.95833333
 1.         0.91666667 0.95833333]


In [18]:
print("Average Accuracy:",scores.mean())
print("Std:",scores.std())
#std of < 0.05 is good.

Average Accuracy: 0.9726666666666668
Std: 0.03212706255087987


## The Avarage accuracy Score of RandomForest model with Cross_val is 0.9726667

Please prefer other notebook for different model